### ISSUE TITLE
@babel/plugin-proposal-class properties: `.name` incorrect for private fields

### ISSUE THREAD

**Issue by mheiber** [babel/babel#10175]
## Bug Report

**Current Behavior**

When a function is assigned to a private field, its `name` property should (I think) be name of the private field.  Instead, the name is always `value`.

**Input Code**

```js
class A {
    #foo = function () {};
    constructor () {
        console.log(this.#foo.name);
    }
}

new A(); // should log 'foo'
```

**Expected behavior/code**

The code above should log 'foo'.

**Babel Configuration (.babelrc, package.json, cli command)**

[repl link](https://babeljs.io/repl#?babili=false&browsers=&build=&builtIns=false&spec=false&loose=false&code_lz=FAYwNghgzlAECCsDexZtgYgGYHsewF5YsBXAOxABcBLHM2ACgEpkBfAblXRDqkoCcSVHP0YsU6SbB5koOMAFMAdGBwBzBpQAW1KEux4lZCAFsFTTpNbBrwMgoDuCZu1gB6N7ChacJMABNYVTVYAHJcHFCgA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=false&presets=es2015%2Ctypescript&prettier=false&targets=&version=7.5.1&externalPlugins=%40babel%2Fplugin-proposal-class-properties%407.5.0)

**Possible Solution**

The function generated WeakMap entry could use the name of the private field ('foo' in this case instead of 'value'). per the [spec proposal](https://tc39.es/proposal-class-fields/#sec-setfunctionname)

```js
  _foo.set(this, {
    writable: true,
    value: function value() {}
  });
```

> Side note: not sure what the 'writable' part is doing: what's a read-only private field?

**Additional context/Screenshots**

- [spec proposal](https://tc39.es/proposal-class-fields/#sec-setfunctionname)

**Comment by JLHwung** [comment]
It seems that V8 will print `#foo` as the function name instead of `foo`.
```js
class A {
    #foo = function () {};
    constructor () {
        console.log(this.#foo.name, this.#foo.name.length);
    }
}

new A();
```
prints `#foo 4` instead of `foo 3` according to the spec.

I have submitted a bug report to V8: https://bugs.chromium.org/p/v8/issues/detail?id=9481

**Comment by mheiber** [comment]
Hi @JLHwung whoops! I think I made a mistake earlier. Sorry about that.

It looks to me now that the function.name should be `#foo`:

https://tc39.es/proposal-class-fields/#sec-runtime-semantics-evaluate-name

lexical semantics:
- privateIdentifier

[runtime semantics](https://tc39.es/proposal-class-fields/#sec-runtime-semantics-evaluate-name)
names are created with newPrivateName(privateIdentifier)

in the [definition of `newPrivateName(description)`](https://tc39.es/proposal-class-fields/#sec-newprivatename):
    -  [[description]] is set to description

in the [lexical grammar](https://tc39.es/proposal-class-fields/#prod-PrivateIdentifier):
- privateIdentifier :: # identifierName

**Comment by mheiber** [comment]
Does this seem to be a real issue, then? I could make a PR in a few weeks after I get approval from my company to contribute.

**Comment by JLHwung** [comment]
@mheiber It is a bug and PR is definitely appreciated.

Note that Babel has similar issues on keeping track of `.name` property, (i.e. #7425 and #9562). Most of this issue can be solved by guess a name from AST and use it as the BindingIdentifier name of the function declaration. 

However, as `#` is not a valid character for identifier name. This one will be more difficult to tackle with than those I mention above. I suggest we may add a `helpers.setFunctionName(f, name)` method, which roughly makes `.name` property writable, writes `name` and locks again.

**Comment by nicolo-ribaudo** [comment]
I would prefer not to use `defineProperty` and to give the name without the hash. Function names are one of the parts of the spec where we aren't really compliant, since what matters most is that they are usable for debugging.

**Comment by JLHwung** [comment]
> Function names are one of the parts of the spec where we aren't really compliant, since what matters most is that they are usable for debugging.

I agree, at least it will be better than our current situation. Though I am afraid it might confuse developers if we return the name _without_ hash while native implementations return the name with hash, which is spec compliant. Maybe we can leave a not comment on that.
